tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/SignificantIndentationSuite.scala

```diff
@@ -0,0 +1,363 @@
+package scala.meta.tests.parsers.dotty
+
+import scala.meta.tests.parsers._
+import scala.meta._
+
+class SignificantIndentationSuite extends BaseDottySuite {
+  implicit val parseStat: String => Stat = code => templStat(code)(dialects.Dotty)
+  implicit val parseSource: String => Source = code => source(code)(dialects.Dotty)
+
+  val defx = Decl.Def(Nil, Term.Name("f"), Nil, Nil, Type.Name("Int"))
+  val defy = Defn.Def(
+    Nil,
+    Term.Name("y"),
+    Nil,
+    Nil,
+    Some(Type.Name("String")),
+    Term.Block(List(Term.Apply(Term.Name("fa"), Nil), Term.Apply(Term.Name("fb"), Nil)))
+  )
+
+  test("basic-example") {
+    val code = """|trait A:
+                  |  def f: Int
+                  |""".stripMargin
+    runTestAssert[Stat](code, assertLayout = Some("trait A { def f: Int }"))(
+      Defn.Trait(
+        Nil,
+        Type.Name("A"),
+        Nil,
+        Ctor.Primary(Nil, Name(""), Nil),
+        Template(Nil, Nil, Self(Name(""), None), List(defx))
+      )
+    )
+  }
+
+  test("multistat-example") {
+    val code = """|trait A:
+                  |  def f: Int
+                  |  def y: String = { fa(); fb() }
+                  |""".stripMargin
+    runTestAssert[Stat](code, assertLayout = None)(
+      Defn.Trait(
+        Nil,
+        Type.Name("A"),
+        Nil,
+        Ctor.Primary(Nil, Name(""), Nil),
+        Template(
+          Nil,
+          Nil,
+          Self(Name(""), None),
+          List(
+            defx,
+            defy
+          )
+        )
+      )
+    )
+  }
+
+  test("indent-and-back") {
+    val code = """|object O:
+                  |  class C:
+                  |    def f: Int = 3
+                  |  trait T:
+                  |    def f: Int = 4
+                  |""".stripMargin
+    runTestAssert[Stat](code, assertLayout = None)(
+      Defn.Object(
+        Nil,
+        Term.Name("O"),
+        Template(
+          Nil,
+          Nil,
+          Self(Name(""), None),
+          List(
+            Defn.Class(
+              Nil,
+              Type.Name("C"),
+              Nil,
+              ctor,
+              Template(
+                Nil,
+                Nil,
+                Self(Name(""), None),
+                List(Defn.Def(Nil, Term.Name("f"), Nil, Nil, Some(Type.Name("Int")), Lit.Int(3)))
+              )
+            ),
+            Defn.Trait(
+              Nil,
+              Type.Name("T"),
+              Nil,
+              Ctor.Primary(Nil, Name(""), Nil),
+              Template(
+                Nil,
+                Nil,
+                Self(Name(""), None),
+                List(Defn.Def(Nil, Term.Name("f"), Nil, Nil, Some(Type.Name("Int")), Lit.Int(4)))
+              )
+            )
+          )
+        )
+      )
+    )
+  }
+
+  test("indent-below-not-okay") {
+    val code = """|def fn: Unit =
```

Comment:
This works in Dotty and apparently it's fine :/

Where is `"error: illegal start of simple expression"` coming from?

Reply:
```scala
def fn: Unit = // next line insert <INDENT=4>
       if cond then   // next indent=2 which is < 4, NO indent
  truep  //indent=2 which is less than previous 4, OUTDENT
       else
```

Reply:
Let's add a comment with a link to an issue: https://github.com/lampepfl/dotty/issues/9790

Reply:
:+1: